qiskit/utils/units.py

```diff
@@ -25,16 +31,113 @@ def apply_prefix(value: float, unit: str) -> float:
     Returns:
         Converted value.
 
+    .. note::
+
+        This may induce tiny value error due to internal representation of float object.
+        See https://docs.python.org/3/tutorial/floatingpoint.html for details.
+
     Raises:
-        Exception: If the units aren't recognized.
+        ValueError: If the ``units`` aren't recognized.
     """
-    downfactors = {"p": 1e12, "n": 1e9, "u": 1e6, "µ": 1e6, "m": 1e3}
-    upfactors = {"k": 1e3, "M": 1e6, "G": 1e9}
-    if not unit:
+    prefactors = {
+        "f": -15,
+        "p": -12,
+        "n": -9,
+        "u": -6,
+        "µ": -6,
+        "m": -3,
+        "k": 3,
+        "M": 6,
+        "G": 9,
+        "T": 12,
+        "P": 15,
+    }
+
+    if not unit or len(unit) == 1:
+        # for example, "m" can represent meter
         return value
-    if unit[0] in downfactors:
-        return value / downfactors[unit[0]]
-    elif unit[0] in upfactors:
-        return value * upfactors[unit[0]]
+
+    if unit[0] not in prefactors:
+        raise ValueError(f"Could not understand unit: {unit}")
+
+    pow10 = prefactors[unit[0]]
+
+    # to avoid round-off error of prefactor
+    if pow10 < 0:
+        return value / pow(10, -pow10)
+
+    return value * pow(10, pow10)
+
+
+def detach_prefix(value: float, decimal: Optional[int] = None) -> Tuple[float, str]:
+    """
+    Given a SI unit value, find the most suitable prefix to scale the value.
+
+    For example, the ``value = 1.3e8`` will be converted into a tuple of ``(130.0, "M")``,
+    which represents a scaled value and auxiliary unit that may be used to display the value.
+    In above example, that value might be displayed as ``130 MHz`` (unit is arbitrary here).
+
+    Example:
+
+        >>> value, prefix = detach_prefix(1e4)
+        >>> print(f"{value} {prefix}Hz")
+        10 kHz
+
+    Args:
+        value: The number to find prefix.
+        decimal: Optional. An arbitrary integer number to represent a precision of the value.
+            If specified, it tries to round the mantissa and adjust the prefix to rounded value.
+            For example, 999_999.91 will become 999.9999 k with ``decimal=4``,
+            while 1.0 M with ``decimal=3`` or less.
+
+    Returns:
+        A tuple of scaled value and prefix.
+
+    Note:
+        This may induce tiny value error due to internal representation of float object.
+        See https://docs.python.org/3/tutorial/floatingpoint.html for details.
```

Comment:
```suggestion
    .. note::

        This may induce tiny value error due to internal representation of float object.
        See https://docs.python.org/3/tutorial/floatingpoint.html for details.
```